### ISSUE TITLE
trigger reportUnnecessaryComparision for "x is None" when "isinstance(x, type(None))" is False

### ISSUE THREAD

**Issue by eamartin** [microsoft/pyright#10141]
**Is your feature request related to a problem? Please describe.**
Pyright will report UnnecessaryIsInstance on `isinstance(x, type(None))` but it doesn't report UnnecessaryComparison on `x is None`.  
I believe when `isinstance(x, type(None))` resolves statically to False then `x is None` can also be statically resolved to False.

This would be a very useful feature. Without this feature, you get better type checks if you replace all instances of `x is None` with `isinstance(x, type(None))` in your codebase.

**Code or Screenshots**

```python
# pyright: strict

def g(x: str):
    if x is None:  # triggers reportUnncessaryComparison
        pass
    if isinstance(x, type(None)):  # triggers reportUnnecesaryIsInstance
        pass

class MyStr(str):
    pass

def f(x: MyStr):
    if x is None:  # doesn't trigger anything !!
        pass
    if isinstance(x, type(None)):  # triggers reportUnnecesaryIsInstance
        pass
```

This outputs
```
test.py
  test.py:4:8 - error: Condition will always evaluate to False since the types "str" and "None" have no overlap (reportUnnecessaryComparison)
  test.py:6:8 - error: Unnecessary isinstance call; "str" is never an instance of "None" (reportUnnecessaryIsInstance)
  test.py:16:8 - error: Unnecessary isinstance call; "MyStr" is never an instance of "None" (reportUnnecessaryIsInstance)
3 errors, 0 warnings, 0 informations
```

**VS Code extension or command-line**
CLI 1.1.397

**Comment by erictraut** [comment]
The `reportUnnecessaryComparision` check was designed to operate on comparison operators (`==` and `!=`), hence its name. It was later extended to support the `is` and `is not` operators using the same logic. Comparison operators can be overridden in python, so type checkers need to be conservative when making assumptions about comparability of two values. For that reason, the logic limits the check to certain known types like `int` and `str` where comparison semantics are well known. With the `is` operator, a type checker can be less conservative, but it is still somewhat limited in what it can assume because types that may initially seem to not overlap might actually overlap via multiple inheritance. The case of `None` on the RHS is special because `NoneType` cannot be subclassed and `None` is a singleton.

In short, this could be implemented in a sound manner. I think it's a reasonable enhancement request. I'll investigate further. If it's straightforward to implement, I'll do so. If it turns out to be a significant investment or present serious risk of regression, I may reject it for now and wait for signal from other pyright users that it's something they'd like to see.

**Comment by erictraut** [comment]
Turned out to be straightforward to implement. This will be included in the next release.

**Comment by eamartin** [comment]
Amazing, thank you!

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.398.
